ffi/src/lib.rs

```diff
@@ -897,3 +897,60 @@ impl<T> Default for ReferenceSet<T> {
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[no_mangle]
+    extern "C" fn allocate_err(etype: KernelError, _: KernelStringSlice) -> *mut EngineError {
+        let boxed = Box::new(EngineError { etype });
+        Box::leak(boxed)
+    }
+
+    fn ok_or_panic<T>(result: ExternResult<T>) -> T {
+        match result {
+            ExternResult::Ok(t) => t,
+            ExternResult::Err(e) => unsafe {
+                panic!("Got engine error with type {:?}", (*e).etype);
+            },
+        }
+    }
+
+    #[test]
+    fn string_slice() {
+        let s = "foo";
+        let _ = kernel_string_slice!(s);
+    }
+
+    #[test]
+    fn bool_slice() {
+        let bools = vec![true, false, true];
+        let bool_slice = KernelBoolSlice::from(bools);
+        unsafe {
+            free_bool_slice(bool_slice);
+        }
+    }
+
+    #[test]
+    fn engine_builder() {
+        let path = "s3://doesntmatter/foo";
+        let path = kernel_string_slice!(path);
+        let builder = unsafe { ok_or_panic(get_engine_builder(path, allocate_err)) };
+        // TODO: When miri supports epoll_wait
+        // let engine = unsafe { builder_build(builder) };
+
+        // for now just rebox so it gets dropped and miri doesn't complain
```

Comment:
need to think if we should return a handle for the builder. not sure why we currently do not.

Reply:
General rule is: The type needs a handle _unless_ it's a public "plain old data" (from C perspective). That includes opaque types, heap-allocated, traits, etc.

Reply:
Right, yeah. I guess with the builder being purely opaque we don't need a handle.